README.md
Google Play Scraper
===================

Fork of https://github.com/undefinedor/google-play-scraper merged with https://github.com/raulr/google-play-scraper 

[![Build Status](https://travis-ci.org/red-eagle/google-play-scraper.svg?branch=master)](https://travis-ci.org/red-eagle/google-play-scraper)

A PHP scraper to get app data from Google Play.

Installation
------------

Add `red-eagle/google-play-scraper` as a require dependency in your `composer.json` file:

```sh
$ composer require red-eagle/google-play-scraper
```

Usage
-----

First create a `Scraper` instance.

```php
use Raulr\GooglePlayScraper\Scraper;

$scraper = new Scraper();
```

There are several methods to configure the default behavior:

* `setDelay($delay)`: Sets the delay in milliseconds between requests to Google Play site.
* `setDefaultLang($lang)`: Sets the default language for all requests. `$lang` must be an [ISO_639-1](https://en.wikipedia.org/wiki/ISO_639-1) two letter language code. If not set, the default language is `en`.
* `setDefaultCountry($country)`: Sets the default country for all requests. `$country` must be an [ISO_3166-1](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2) two letter country code. If not set, the default country is `us`.

### getApp

Gets app information given its ID.

#### Parameters

* `$id`: Google Play app identifier.
* `$lang`: (optional, defaults to `null`): Overrides the default language.
* `$country`: (optional, defaults to `null`): Overrides the default country.

#### Example

```php
$app = $scraper->getApp('com.mojang.minecraftpe');
```

Returns:

```php
array (
  'id' => 'com.mojang.minecraftpe',
  'url' => 'https://play.google.com/store/apps/details?id=com.mojang.minecraftpe',
  'image' => 'https://lh3.googleusercontent.com/30koN0eGl-LHqvUZrCj9HT4qVPQdvN508p2wuhaWUnqKeCp6nrs9QW8v6IVGvGNauA=w300',
  'title' => 'Minecraft: Pocket Edition',
  'author' => 'Mojang',
  'author_link' => 'https://play.google.com/store/apps/developer?id=Mojang',
  'categories' => array (
    'Arcade',
    'Creativity',
  ),
  'price' => '$6.99',
  'screenshots' => array (
    'https://lh3.googleusercontent.com/VkLE0e0EDuRID6jdTE97cC8BomcDReJtZOem9Jlb14jw9O7ytAGvE-2pLqvoSJ7w3IdK=h310',
    'https://lh3.googleusercontent.com/28b1vxJQe916wOaSVB4CmcnDujk8M2SNaCwqtQ4cUS0wYKYn9kCYeqxX0uyI2X-nQv0=h310',
    // [...]
  ),
  'description' => 'Our latest free update includes the Nether and all its inhabitants[...]',
  'description_html' => 'Our latest free update includes the Nether and all its inhabitants[...]',
  'rating' => 4.4726405143737793,
  'votes' => 1136962,
  'last_updated' => 'October 22, 2015',
  'size' => 'Varies with device',
  'downloads' => '10,000,000 - 50,000,000',
  'version' => 'Varies with device',
  'supported_os' => 'Varies with device',
  'content_rating' => 'Everyone 10+',
  'whatsnew' => 'Build, explore and survive on the go with Minecraft: Pocket Edition[...]',
  'video_link' => 'https://www.youtube.com/embed/D2Z9oKTzzrM?ps=play&vq=large&rel=0&autohide=1&showinfo=0&autoplay=1',
  'video_image' => 'https://i.ytimg.com/vi/D2Z9oKTzzrM/hqdefault.jpg',
)
```

The following fields may contain a `null` value: `price`, `size`, `downloads`, `version`, `whatsnew`, `video_link` and `video_image`. The `price` being `null` means the app is free.

### getApps

Gets information for multiple apps given their IDs.

#### Parameters

* `$ids`: Array of Google Play app identifiers.
* `$lang`: (optional, defaults to `null`): Overrides the default language.
* `$country`: (optional, defaults to `null`): Overrides the default country.

#### Example

```php
$app = $scraper->getApps(array(
    'com.mojang.minecraftpe',
    'com.google.android.youtube',
));
```

### getCategories

Returns an array with the existing categories in Google Play.

#### Example

```php
use Raulr\GooglePlayScraper\Scraper;

$scraper = new Scraper();
$categories = $scraper->getCategories();
```
Returns:

```php
array (
  'BOOKS_AND_REFERENCE',
  'BUSINESS',
  'COMICS',
  'COMMUNICATION',
  'EDUCATION',
  'ENTERTAINMENT',
  'FINANCE',
  'HEALTH_AND_FITNESS',
  'LIBRARIES_AND_DEMO',
  'LIFESTYLE',
  'APP_WALLPAPER',
  'MEDIA_AND_VIDEO',
  'MEDICAL',
  'MUSIC_AND_AUDIO',
  'NEWS_AND_MAGAZINES',
  'PERSONALIZATION',
  'PHOTOGRAPHY',
  'PRODUCTIVITY',
  'SHOPPING',
  'SOCIAL',
  'SPORTS',
  'TOOLS',
  'TRANSPORTATION',
  'TRAVEL_AND_LOCAL',
  'WEATHER',
  'APP_WIDGETS',
  'GAME_ACTION',
  'GAME_ADVENTURE',
  'GAME_ARCADE',
  'GAME_BOARD',
  'GAME_CARD',
  'GAME_CASINO',
  'GAME_CASUAL',
  'GAME_EDUCATIONAL',
  'GAME_MUSIC',
  'GAME_PUZZLE',
  'GAME_RACING',
  'GAME_ROLE_PLAYING',
  'GAME_SIMULATION',
  'GAME_SPORTS',
  'GAME_STRATEGY',
  'GAME_TRIVIA',
  'GAME_WORD',
  'FAMILY',
  'FAMILY_ACTION',
  'FAMILY_BRAINGAMES',
  'FAMILY_CREATE',
  'FAMILY_EDUCATION',
  'FAMILY_MUSICVIDEO',
  'FAMILY_PRETEND',
)
```

### getCollections

Returns an array with the existing collections in Google Play.

#### Example

```php
$collections = $scraper->getCollections();
```
Returns:

```php
array (
  'topselling_free',
  'topselling_paid',
  'topselling_new_free',
  'topselling_new_paid',
  'topgrossing',
  'movers_shakers',
)
```

### getList

Retrieves a list of Google Play apps given a collection and optionally filtered by category.

#### Parameters

* `$collection`: Google Play collection to retrieve. See [getCollections](#getcollections) for possible values.
* `$category`: (optional, defaults to `null`) Filter request by this category. See [getCategories](#getcategories) for possible values.
* `$lang`: (optional, defaults to `null`): Overrides the default language.
* `$country`: (optional, defaults to `null`): Overrides the default country.

#### Example

```php
$apps = $scraper->getList('topselling_free', 'SOCIAL');
```

Returns:

```php
array (
  array (
    'id' => 'com.facebook.katana',
    'url' => 'https://play.google.com/store/apps/details?id=com.facebook.katana',
    'title' => 'Facebook',
    'image' => 'https://lh3.googleusercontent.com/ZZPdzvlpK9r_Df9C3M7j1rNRi7hhHRvPhlklJ3lfi5jk86Jd1s0Y5wcQ1QgbVaAP5Q=w340',
    'author' => 'Facebook',
    'rating' => 3.9888803958892822,
    'price' => null,
  ),
  array (
    'id' => 'com.snapchat.android',
    'url' => 'https://play.google.com/store/apps/details?id=com.snapchat.android',
    'title' => 'Snapchat',
    'image' => 'https://lh4.ggpht.com/vdK_CsMSsJoYvJpYgaj91fiJ1T8rnSHHbXL0Em378kQaaf_BGyvUek2aU9z2qbxJCAFV=w340',
    'author' => 'Snapchat Inc',
    'rating' => 3.8660063743591309,
    'price' => null,
  ),
  // [...]
)
```

### getDetailList

Same as [getList](#getlist) but returning full detail app data. An additional request is made for every app from the list in order to get its details.

### getListChunk

Retrieves a chunk of a Google Play app list.

#### Parameters

* `$collection`: Google Play collection to retrieve. See [getCollections](#getcollections) for possible values.
* `$category`: (optional, defaults to `null`) Filter request by this category. See [getCategories](#getcategories) for possible values.
* `$start`: (optional, defaults to `0`): Starting index. Must be a value between `0` and `500`.
* `$num`: (optional, defaults to `60`): Amount of apps to retrieve. Must be a value between `0` and `120`.
* `$lang`: (optional, defaults to `null`): Overrides the default language.
* `$country`: (optional, defaults to `null`): Overrides the default country.

#### Example

```php
$apps = $scraper->getListChunk('topselling_free', 'SOCIAL', 20, 80);
```

### getDetailListChunk

Same as [getListChunk](#getlistchunk) but returning full detail app data. An additional request is made for every app from the list in order to get its details.

### getSearch

Retrieves a list of Google Play apps given a search query and optionally filtered by price and rating.

#### Parameters

* `$query`: Search query.
* `$price`: (optional, defaults to `all`) Filter request by price. Possible values: `all`, `free`, `paid`.
* `$rating`: (optional, defaults to `all`) Filter request by rating. Possible values: `all`, `4+`.
* `$lang`: (optional, defaults to `null`): Overrides the default language.
* `$country`: (optional, defaults to `null`): Overrides the default country.

#### Example

```php
$apps = $scraper->getSearch('unicorns', 'free', '4+');
```

### getDetailSearch

Same as [getSearch](#getsearch) but returning full detail app data. An additional request is made for every app from the search result in order to get its details.

### getCommentCount

Retrieves comment count for application on selected language

#### Parameters

* `$id` - application google play id
* `$lang` - language

#### Example

```php
$commentCount = $scraper->getCommentCount('com.facebook.katana', 'en');
```

### getAllLangCommentCount

Same as [getCommentCount](#getcommentcount) but retrieves comment count for all selected language (array as second paramter)

#### Parameters

* `$id` - application google play id
* `$langs` - languages (array)

### getSimilar

Retrieves similar apps

#### Parameters

* `$id` - application google play id

### getDetailSimilar

Same as [getDetailSimilar](#getdetailsimilar) but retrieves apps with details

* `$id` - application google play id

CHANGELOG.md
# CHANGELOG

## 0.2.1 (2017-01-08)

### Added

* `getAllLanguages`, `getCommentCount` and `getAllLangCommentCount` methods.

## 0.2.0 (2016-11-10)

### Added

* `getSearch` and `getDetailSearch` methods.

### Changed

* Convert schemaless URLs into absolute URLs.

## 0.1.1 (2016-04-09)

### Fixed

* Fix throwing InvalidArgumentException on `Scraper::getApp()` when app page contains a `NULL` byte.
* Remove duplicate license field on composer.json.

## 0.1.0 (2016-01-04)

* Initial release.


src/Scraper.php
<?php

namespace Raulr\GooglePlayScraper;

use Goutte\Client;
use GuzzleHttp\ClientInterface as GuzzleClientInterface;
use Raulr\GooglePlayScraper\Exception\BannedException;
use Symfony\Component\DomCrawler\Crawler;
use Raulr\GooglePlayScraper\Exception\RequestException;
use Raulr\GooglePlayScraper\Exception\NotFoundException;

/**
 * @author Raul Rodriguez <raul@raulr.net>
 */
class Scraper
{
    const BASE_URL = 'https://play.google.com';

    protected $client;
    protected $delay = 1000;
    protected $lastRequestTime;
    protected $lang = 'en';
    protected $country = 'us';
    protected $debug = false;

    public function __construct(GuzzleClientInterface $guzzleClient = null)
    {
        $this->client = new Client();
        if ($guzzleClient) {
            $this->client->setClient($guzzleClient);
        }
    }

    public function setDelay($delay)
    {
        $this->delay = intval($delay);
    }

    public function getDelay()
    {
        return $this->delay;
    }

    public function setDefaultLang($lang)
    {
        $this->lang = $lang;
    }

    public function getDefaultLang()
    {
        return $this->lang;
    }

    public function setDefaultCountry($country)
    {
        $this->country = $country;
    }

    public function getDefaultCountry()
    {
        return $this->country;
    }

    public function setDebug($val)
    {
        $this->debug = (bool)$val;
    }

    public function getDebug()
    {
        return $this->debug;
    }

    public function getCategories()
    {
        $crawler = $this->request('apps', array(
            'hl' => 'en',
            'gl' => 'us',
        ));

        $collections = $crawler
            ->filter('.child-submenu-link')
            ->reduce(function ($node) {
                return strpos($node->attr('href'), '/store/apps') === 0;
            })
            ->each(function ($node) {
                $href = $node->attr('href');
                $hrefParts = explode('/', $href);
                $collection = end($hrefParts);
                $collection = preg_replace('/\?.*$/', '', $collection);

                return $collection;
            });
        $collections = array_unique($collections);

        return $collections;
    }

    public function getCollections()
    {
        return array(
            'topselling_free',
            'topselling_paid',
            'topselling_new_free',
            'topselling_new_paid',
            'topgrossing',
            'movers_shakers',
        );
    }

    public function getApp($id, $lang = null, $country = null)
    {
        $lang = $lang === null ? $this->lang : $lang;
        $country = $country === null ? $this->country : $country;

        $params = array(
            'id' => $id,
            'hl' => $lang,
            'gl' => $country,
        );
        $crawler = $this->request(array('apps', 'details'), $params);

        $info = array();
        $info['id'] = $id;
        $info['url'] = $crawler->filter('[itemprop="url"]')->attr('content');
        $info['image'] = $this->getAbsoluteUrl($crawler->filter('[itemprop="image"]')->attr('src'));
        $info['title'] = $crawler->filter('[itemprop="name"] > div')->text();
        $info['author'] = $crawler->filter('[itemprop="author"] [itemprop="name"]')->text();
        $info['author_link'] = $this->getAbsoluteUrl($crawler->filter('[itemprop="author"] > [itemprop="url"]')->attr('content'));
        $info['categories'] = $crawler->filter('[itemprop="genre"]')->each(function ($node) {
            return $node->text();
        });
        $price = $crawler->filter('[itemprop="offers"] > [itemprop="price"]')->attr('content');
        $info['price'] = $price == '0' ? null : $price;
        $info['screenshots'] = $crawler->filter('[itemprop="screenshot"]')->each(function ($node) {
            return $this->getAbsoluteUrl($node->filter('img')->attr('src'));
        });
        $desc = $this->cleanDescription($crawler->filter('[itemprop="description"] > div'));
        $info['description'] = $desc['text'];
        $info['description_html'] = $desc['html'];
        $ratingNode = $crawler->filter('[itemprop="aggregateRating"] > [itemprop="ratingValue"]');
        if ($ratingNode->count()) {
            $rating = floatval($ratingNode->attr('content'));
        } else {
            $rating = 0.0;
        }
        $info['rating'] = $rating;
        $votesNode = $crawler->filter('[itemprop="aggregateRating"] > [itemprop="ratingCount"]');
        if ($votesNode->count()) {
            $votes = intval($votesNode->attr('content'));
        } else {
            $votes = 0;
        }
        $info['votes'] = $votes;
        $info['last_updated'] = trim($crawler->filter('[itemprop="datePublished"]')->text());
        $sizeNode = $crawler->filter('[itemprop="fileSize"]');
        if ($sizeNode->count()) {
            $size = trim($sizeNode->text());
        } else {
            $size = null;
        }
        $info['size'] = $size;
        $downloadsNode = $crawler->filter('[itemprop="numDownloads"]');
        if ($downloadsNode->count()) {
            $downloads = trim($downloadsNode->text());
        } else {
            $downloads = null;
        }
        $info['downloads'] = $downloads;
        $versionNode = $crawler->filter('[itemprop="softwareVersion"]');
        if ($versionNode->count()) {
            $version = trim($versionNode->text());
        } else {
            $version = null;
        }
        $info['version'] = $version;
        $info['supported_os'] = trim($crawler->filter('[itemprop="operatingSystems"]')->text());
        $info['content_rating'] = $crawler->filter('[itemprop="contentRating"]')->text();
        $whatsneNode = $crawler->filter('.recent-change');
        if ($whatsneNode->count()) {
            $info['whatsnew'] = implode("\n", $whatsneNode->each(function ($node) {
                return $node->text();
            }));
        } else {
            $info['whatsnew'] = null;
        }
        $videoNode = $crawler->filter('.details-trailer');
        if ($videoNode->count()) {
            $info['video_link'] = $this->getAbsoluteUrl($videoNode->filter('.play-action-container')->attr('data-video-url'));
            $info['video_image'] = $this->getAbsoluteUrl($videoNode->filter('.video-image')->attr('src'));
        } else {
            $info['video_link'] = null;
            $info['video_image'] = null;
        }

        // Comment count
        //$info['comment_count'] = $this->getCommentCount($id, $lang);
        //All languages
        $info['all_languages'] = $this->_getLanguages();

        return $info;
    }

    public function getApps($ids, $lang = null, $country = null)
    {
        $ids = (array)$ids;
        $apps = array();

        foreach ($ids as $id) {
            $apps[$id] = $this->getApp($id, $lang, $country);
        }

        return $apps;
    }

    public function getListChunk($collection, $category = null, $start = 0, $num = 60, $lang = null, $country = null)
    {
        $lang = $lang === null ? $this->lang : $lang;
        $country = $country === null ? $this->country : $country;

        if (!is_int($start)) {
            throw new \InvalidArgumentException('"start" must be an integer');
        }
        if ($start < 0 || $start > 500) {
            throw new \RangeException('"start" must be a number between 0 and 500');
        }
        if (!is_int($num)) {
            throw new \InvalidArgumentException('"num" must be an integer');
        }
        if ($num < 0 || $num > 120) {
            throw new \RangeException('"num" must be a number between 0 and 120');
        }

        $path = array('apps');
        if ($category) {
            array_push($path, 'category', $category);
        }
        array_push($path, 'collection', $collection);
        $params = array(
            'hl' => $lang,
            'gl' => $country,
            'start' => $start,
            'num' => $num,
        );
        $crawler = $this->request($path, $params);

        return $this->parseAppList($crawler);
    }

    public function getList($collection, $category = null, $lang = null, $country = null)
    {
        $lang = $lang === null ? $this->lang : $lang;
        $country = $country === null ? $this->country : $country;
        $start = 0;
        $num = 60;
        $apps = array();
        $appsChunk = array();

        do {
            $appsChunk = $this->getListChunk($collection, $category, $start, $num, $lang, $country);
            $apps = array_merge($apps, $appsChunk);
            $start += $num;
        } while (count($appsChunk) == $num && $start <= 500);

        return $apps;
    }

    public function getDetailListChunk(
        $collection,
        $category = null,
        $start = 0,
        $num = 60,
        $lang = null,
        $country = null
    ) {
        $apps = $this->getListChunk($collection, $category, $start, $num, $lang, $country);
        $ids = array_map(function ($app) {
            return $app['id'];
        }, $apps);

        return $this->getApps($ids);
    }

    public function getDetailList($collection, $category = null, $lang = null, $country = null)
    {
        $apps = $this->getList($collection, $category, $lang, $country);
        $ids = array_map(function ($app) {
            return $app['id'];
        }, $apps);

        return $this->getApps($ids);
    }

    public function getSearch($query, $price = 'all', $rating = 'all', $lang = null, $country = null)
    {
        $lang = $lang === null ? $this->lang : $lang;
        $country = $country === null ? $this->country : $country;
        $priceValues = array(
            'all' => null,
            'free' => 1,
            'paid' => 2,
        );
        $ratingValues = array(
            'all' => null,
            '4+' => 1,
        );

        if (!is_string($query) || empty($query)) {
            throw new \InvalidArgumentException('"query" must be a non empty string');
        }

        if (array_key_exists($price, $priceValues)) {
            $price = $priceValues[$price];
        } else {
            throw new \InvalidArgumentException('"price" must contain one of the following values: ' . implode(', ',
                    array_keys($priceValues)));
        }

        if (array_key_exists($rating, $ratingValues)) {
            $rating = $ratingValues[$rating];
        } else {
            throw new \InvalidArgumentException('"rating" must contain one of the following values: ' . implode(', ',
                    array_keys($ratingValues)));
        }

        $apps = array();
        $path = array('search');
        $params = array(
            'q' => $query,
            'c' => 'apps',
            'hl' => $lang,
            'gl' => $country,
        );
        if ($price) {
            $params['price'] = $price;
        }
        if ($rating) {
            $params['rating'] = $rating;
        }

        do {
            $crawler = $this->request($path, $params);
            $apps = array_merge($apps, $this->parseAppList($crawler));
            unset($params['pagTok']);
            foreach ($crawler->filter('script') as $scriptNode) {
                if (preg_match('/\\\x22(GAE.+?)\\\x22/', $scriptNode->textContent, $matches)) {
                    $params['pagTok'] = preg_replace('/\\\\\\\u003d/', '=', $matches[1]);
                    break;
                }
            }
        } while (array_key_exists('pagTok', $params));

        return $apps;
    }

    public function getDetailSearch($query, $price = 'all', $rating = 'all', $lang = null, $country = null)
    {
        $apps = $this->getSearch($query, $price, $rating, $lang, $country);
        $ids = array_map(function ($app) {
            return $app['id'];
        }, $apps);

        return $this->getApps($ids);
    }

    protected function request($path, array $params = array())
    {
        // handle delay
        $this->_handleDelay();

        $this->lastRequestTime = microtime(true);

        if (is_array($path)) {
            $path = implode('/', $path);
        }
        $path = ltrim($path, '/');
        $path = rtrim('/store/' . $path, '/');
        $url = self::BASE_URL . $path;
        $query = http_build_query($params);
        if ($query) {
            $url .= '?' . $query;
        }
        $crawler = $this->client->request('GET', $url);
        $status_code = $this->client->getResponse()->getStatus();
        if ($status_code == 404) {
            throw new NotFoundException('Requested resource not found');
        } elseif ($status_code != 200) {
            throw new RequestException(sprintf('Request failed with "%d" status code', $status_code), $status_code);
        }

        return $crawler;
    }

    protected function getAbsoluteUrl($url)
    {
        $urlParts = parse_url($url);
        $baseParts = parse_url(self::BASE_URL);
        $absoluteParts = array_merge($baseParts, $urlParts);

        $absoluteUrl = $absoluteParts['scheme'] . '://' . $absoluteParts['host'];
        if (isset($absoluteParts['path'])) {
            $absoluteUrl .= $absoluteParts['path'];
        } else {
            $absoluteUrl .= '/';
        }
        if (isset($absoluteParts['query'])) {
            $absoluteUrl .= '?' . $absoluteParts['query'];
        }
        if (isset($absoluteParts['fragment'])) {
            $absoluteUrl .= '#' . $absoluteParts['fragment'];
        }

        return $absoluteUrl;
    }

    protected function parseAppList(Crawler $crawler)
    {
        return $crawler->filter('.card')->each(function ($node) {
            $app = array();
            $app['id'] = $node->attr('data-docid');
            $app['url'] = self::BASE_URL . $node->filter('a')->attr('href');
            $app['title'] = $node->filter('a.title')->attr('title');
            $app['image'] = $this->getAbsoluteUrl($node->filter('img.cover-image')->attr('data-cover-large'));
            $app['author'] = $node->filter('a.subtitle')->attr('title');
            $ratingNode = $node->filter('.current-rating');
            if (!$ratingNode->count()) {
                $rating = 0.0;
            } elseif (preg_match('/\d+(\.\d+)?/', $node->filter('.current-rating')->attr('style'), $matches)) {
                $rating = floatval($matches[0]) * 0.05;
            } else {
                throw new \RuntimeException('Error parsing rating');
            }
            $app['rating'] = $rating;
            $priceNode = $node->filter('.display-price');
            if (!$priceNode->count()) {
                $price = null;
            } elseif (!preg_match('/\d/', $priceNode->text())) {
                $price = null;
            } else {
                $price = $priceNode->text();
            }
            $app['price'] = $price;

            return $app;
        });
    }

    protected function cleanDescription(Crawler $descriptionNode)
    {
        $descriptionNode->filter('a')->each(function ($node) {
            $domElement = $node->getNode(0);
            $href = $domElement->getAttribute('href');
            while (strpos($href, 'https://www.google.com/url?q=') === 0) {
                $parts = parse_url($href);
                parse_str($parts['query'], $query);
                $href = $query['q'];
            }
            $domElement->setAttribute('href', $href);
        });
        $html = $descriptionNode->html();
        $text = trim($this->convertHtmlToText($descriptionNode->getNode(0)));

        return array(
            'html' => $html,
            'text' => $text,
        );
    }

    protected function convertHtmlToText(\DOMNode $node)
    {
        if ($node instanceof \DOMText) {
            $text = preg_replace('/\s+/', ' ', $node->wholeText);
        } else {
            $text = '';

            foreach ($node->childNodes as $childNode) {
                $text .= $this->convertHtmlToText($childNode);
            }

            switch ($node->nodeName) {
                case 'h1':
                case 'h2':
                case 'h3':
                case 'h4':
                case 'h5':
                case 'h6':
                case 'p':
                case 'ul':
                case 'div':
                    $text = "\n\n" . $text . "\n\n";
                    break;
                case 'li':
                    $text = '- ' . $text . "\n";
                    break;
                case 'br':
                    $text = $text . "\n";
                    break;
            }

            $text = preg_replace('/\n{3,}/', "\n\n", $text);
        }

        return $text;
    }

    /**
     * Get comment count.
     */
    public function getAllLangCommentCount($id, $langs = [])
    {
        return count($this->getAllLangCommentIds($id, $langs));
    }

    /**
     *  reviewType:0
     *  pageNum:0
     *  id:com.onetongames.realbattlesimulator
     *  reviewSortOrder:0
     *  xhr:1
     *  hl:en.
     */
    public function getCommentCount($id, $lang = null)
    {
        return count($this->getCommentIds($id, $lang));
    }

    /**
     * @return array
     */
    public function getCommentIds($id, $lang = null)
    {
        $params = [
            'reviewType' => 0,
            'pageNum' => 0,
            'id' => $id,
            'reviewSortOrder' => 0,
            'xhr' => 1,
            'hl' => is_null($lang) ? $this->lang : $lang,
        ];
        $comments = [];
        do {
            $this->_handleDelay();
            $this->lastRequestTime = microtime(true);
            try {
                $response = $this->client->getClient()->post(self::BASE_URL . '/store/getreviews', [
                    'form_params' => $params,
                    'allow_redirects' => true,
                ]);
                $text = trim(ltrim($response->getBody()->getContents(), ")]}'"));
                $data = \GuzzleHttp\json_decode($text, true);
            } catch (\Exception $exception) {
                throw new BannedException();
            }
            if (preg_match_all('/(?<=data-reviewid=")gp:[^"]+/', $data[0][2], $reviews, PREG_PATTERN_ORDER)) {
                $comments = array_merge($comments, $reviews[0]);
            } else {
                break;
            }
            ++$params['pageNum'];
        } while (true);
        return $comments;
    }

    public function getAllLangCommentIds($id, $langs = [])
    {
        if (empty($langs)) {
            $lang = [$this->getDefaultLang()];
        }
        $comments = [];
        foreach ($langs as $lang) {
            $comments = array_merge($this->getCommentIds($id, $lang), $comments);
        }

        return array_unique($comments);
    }

    public function getSimilar($id, $lang = null, $country = null)
    {
        $crawler = $this->request('apps/similar', [
            'id' => $id,
            'hl' => is_null($lang) ? $this->getDefaultLang() : $lang,
            'gl' => is_null($country) ? $this->getDefaultCountry() : $country,
        ]);
        return $this->parseAppList($crawler);
    }

    public function getDetailSimilar($id, $lang = null, $country = null) {
        return array_map(function($app) {
            return $this->getApp($app['id']);
        }, $this->getSimilar($id));
    }

    private function _getLanguages()
    {
        $crawler = $this->client->getCrawler();
        $nodes = $crawler->filter('head > link[hreflang]');
        if ($nodes->count() === 0) {
            return [];
        }
        $result = $nodes->each(function ($node) {
            return $node->attr('hreflang');
        });
        array_shift($result); //Delete first element "x-default"
        return $result;
    }

    /**
     * Delay.
     */
    protected function _handleDelay()
    {
        // handle delay
        if (!empty($this->delay) && !empty($this->lastRequestTime)) {
            $currentTime = microtime(true);
            $delaySecs = $this->delay / 1000;
            $delay = max(0, $delaySecs - $currentTime + $this->lastRequestTime);
            usleep($delay * 1000000);
        }
    }
}
